File: 2689-rearranging-fruits/solution.cpp
class Solution {
    static bool processBasket(vector<int>& basket) {
        const int length = basket.size();

        int i, j;
        for (i = 0; (i != length) && (basket[i] != -1); i++) {}
        for (j = i; (j != length) && (basket[j] == -1); j++) {}

        while (j != length) {
            basket[i++] = basket[j++];
            for (; (j != length) && (basket[j] == -1); j++) {}
        }

        basket.resize(i);

        if (i == 0)
            return true;

        int count = 1;

        for (i -= 2; i != -1; i--)
            if (basket[i] != basket[i + 1]) {
                if ((count & 1) == 1)
                    return false;
                
                count = 1;
            } else
                count++;

        return (count & 1) != 1;
    }

public:
    long long minCost(vector<int>& basket1, vector<int>& basket2) {
        sort(basket1.begin(), basket1.end());
        sort(basket2.begin(), basket2.end());

        const int bl = basket1.size(), minCost = min(basket1[0], basket2[0]) * 2;
        
        for (int i1 = 0, i2 = 0; (i1 != bl) && (i2 != bl);)
            if (basket1[i1] < basket2[i2])
                i1++;
            else if (basket2[i2] < basket1[i1])
                i2++;
            else
                basket1[i1++] = basket2[i2++] = -1;
        
        if (!processBasket(basket1) || !processBasket(basket2))
            return -1;

        long long result = 0;

        for (int i1 = 0, i2 = 0, c = basket1.size(); c != 0; c--)
            result += min(minCost,
                basket1[i1] < basket2[i2] ? basket1[i1++] : basket2[i2++]);

        return result / 2;
    }
};

File: 2689-rearranging-fruits/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn min_cost(basket1: Vec<i32>, basket2: Vec<i32>) -> i64 {
        let mut map: HashMap<i32, i32> = HashMap::new();
        let mut min_val = i32::MAX;
        for &x in &basket1 {
            *map.entry(x).or_insert(0) += 1;
            min_val = min_val.min(x);
        }
        for &x in &basket2 {
            *map.entry(x).or_insert(0) -= 1;
            min_val = min_val.min(x);
        }

        let mut to_swap = Vec::new();
        for (&val, &diff) in map.iter() {
            if diff % 2 != 0 { 
                return -1; 
            }
            let count = diff.abs() / 2;
            for _ in 0..count {
                to_swap.push(val);
            }
        }

        to_swap.sort_unstable();
        
        let mut res = 0i64;
        let n = to_swap.len() / 2;
        
        for i in 0..n {
            res += (to_swap[i].min(2 * min_val)) as i64;
        }
        
        res
    }
}
